doc/developers.rst

```diff
@@ -1417,3 +1417,134 @@ Of course, a given field may already be on the device (and have been
+    CALL ocl_env_init
+    ! The kernels this PSy layer module requires
+    kernel_names(1) = "compute_cu_code"
+    ! Create the OpenCL kernel objects. Expects to find all of the
+    ! compiled kernels in PSYCLONE_KERNELS_FILE.
+    CALL add_kernels(1, kernel_names)
+
+As indicated in the comment, the ``FortCL::add_kernels`` routine
+expects to find all kernels in a pre-compiled file pointed to by the
+PSYCLONE_KERNELS_FILE environment variable. (A pre-compiled file is
+used instead of run-time kernel compilation in order to support
+execution on FPGAs.)
+
+The second routine created by PSyclone sets the kernel arguments, e.g.::
+
+    SUBROUTINE compute_cu_code_set_args(kernel_obj, nx, cu_fld, p_fld, u_fld)
+      USE clfortran, ONLY: clSetKernelArg
+      USE iso_c_binding, ONLY: c_sizeof, c_loc, c_intptr_t
+      ...
+      INTEGER(KIND=c_intptr_t), target :: cu_fld, p_fld, u_fld
+      INTEGER(KIND=c_intptr_t), target :: kernel_obj
+      INTEGER, target :: nx
+      ! Set the arguments for the compute_cu_code OpenCL Kernel
+      ierr = clSetKernelArg(kernel_obj, 0, C_SIZEOF(nx), C_LOC(nx))
+      ierr = clSetKernelArg(kernel_obj, 1, C_SIZEOF(cu_fld), C_LOC(cu_fld))
+      ...
+    END SUBROUTINE compute_cu_code_set_args
+
+The third routine generated is the ususal psy-layer routine that is
+responsible for calling all of the kernels. However, it must now also
+call ``psy_init``, create buffers on the compute device (if they are
+not already present) and copy data over::
+
+    SUBROUTINE invoke_compute_cu(...)
+      ...
+      IF (first_time) THEN
+        first_time = .false.
+        CALL psy_init
+        num_cmd_queues = get_num_cmd_queues()
+        cmd_queues => get_cmd_queues()
+        kernel_compute_cu_code = get_kernel_by_name("compute_cu_code")
+      END IF 
+      globalsize = (/p_fld%grid%nx, p_fld%grid%ny/)
+      ! Ensure field data is on device
+      IF (.NOT. cu_fld%data_on_device) THEN
+        size_in_bytes = int(p_fld%grid%nx*p_fld%grid%ny, 8)* &
+                        c_sizeof(cu_fld%data(1,1))
+        ! Create buffer on device
+        cu_fld%device_ptr = create_rw_buffer(size_in_bytes)
+        ierr = clEnqueueWriteBuffer(cmd_queues(1), cu_fld%device_ptr,  &
+                                    CL_TRUE, 0_8, size_in_bytes,       &
+      			            C_LOC(cu_fld%data), 0, C_NULL_PTR, &
+      			            C_LOC(write_event))
+        cu_fld%data_on_device = .true.
+      END IF 
+      ...
+
+Note that we use the ``data_on_device`` member of the field derived
+type (implement in github.com/stfc/dl_esm_inf) to keep track of
+whether a given field has been copied to the compute device.  Once all
+of this setup is done, the kernel itself is launched by calling
+``clEnqueueNDRangeKernel``::
+
+    ierr = clEnqueueNDRangeKernel(cmd_queues(1), kernel_compute_cu_code, &
+                                  2, C_NULL_PTR, C_LOC(globalsize),      &
+				  C_NULL_PTR, 0, C_NULL_PTR, C_NULL_PTR)
+
+Limitations
```

Comment:
Good point. I was originally (fondly) imagining that the OpenCL runtime would allow us to target both the CPU and the FPGA but that's definitely not going to happen in EuroEXA. I can imagine it might be something that Intel will support in future (if they don't already). I've added a couple of lines.